cleverhans/future/torch/attacks/carlini_wagner_l2.py

```diff
@@ -0,0 +1,205 @@
+            array of logits. Default is False.
+    """
+
+    # Convert logits to predicted class if necessary
+    if is_logits:
+      pred_copy = pred.clone().detach()
+      pred_copy[label] += (-confidence if targeted else confidence)
+      pred = torch.argmax(pred_copy)
+
+    return pred == label if targeted else pred != label
+
+
+  if y is None:
+    # Using model predictions as ground truth to avoid label leaking
+    pred = model_fn(x)
+    y = torch.argmax(pred, 1)
+
+  # Initialize some values needed for binary search on const
+  lower_bound = [0.] * len(x)
+  upper_bound = [1e10] * len(x)
+  const = x.new_ones(len(x), 1) * initial_const
+
+  o_bestl2 = [INF] * len(x)
+  o_bestscore = [-1.] * len(x)
+  x = torch.clamp(x, clip_min, clip_max)
+  ox = x.clone().detach() # save the original x
+  o_bestattack = x.clone().detach()
+
+  # Map images into the tanh-space
+  # TODO as of 01/06/2020, PyTorch does not natively support
+  # arctanh (see, e.g.,
+  # https://github.com/pytorch/pytorch/issues/10324).
+  # This particular implementation here is not numerically
+  # stable and should be substituted w/ PyTorch's native
+  # implementation when it comes out in the future
+  arctanh = lambda x: 0.5 * torch.log((1 + x) / (1 - x))
+  x = (x - clip_min) / (clip_max - clip_min)
+  x = torch.clamp(x, 0, 1)
+  x = x * 2 - 1
+  x = arctanh(x * .999999)
+
+  # Prepare some variables
+  modifier = torch.zeros_like(x, requires_grad=True)
+  y_onehot = y.new_zeros((len(y), n_classes), dtype=torch.float)
+  y_onehot[range(len(y)), y] = 1.
+
+  # Define loss functions and optimizer
+  f_fn = lambda real, other, targeted: torch.max(
+    ((other - real) if targeted else (real - other)) + confidence,
+    torch.tensor(0.)
+  )
+  l2dist_fn = lambda x, y: torch.pow(x - y, 2).sum(list(range(len(x.size())))[1:])
+  optimizer = torch.optim.Adam([modifier], lr=lr)
+
+  # Outer loop performing binary search on const
+  for outer_step in range(binary_search_steps):
+    # Initialize some values needed for the inner loop
+    bestl2 = [INF] * len(x)
+    bestscore = [-1.] * len(x)
+
+    # Inner loop performing attack iterations
+    for i in range(max_iterations):
+      # One attack step
+      new_x = (torch.tanh(modifier + x) + 1) / 2
+      new_x = new_x * (clip_max - clip_min) + clip_min
+      logits = model_fn(new_x)
+
+      real = torch.sum(y_onehot * logits, 1)
+      other, _ = torch.max((1 - y_onehot) * logits - y_onehot * 1e4, 1)
+
+      optimizer.zero_grad()
+      f = f_fn(real, other, targeted)
+      l2 = l2dist_fn(new_x, ox)
+      loss = (const * f + l2).sum()
+      loss.backward()
+      optimizer.step()
+
+      # Update best results
+      for n, (l2_n, logits_n, new_x_n) in enumerate(zip(l2, logits, new_x)):
+        y_n = y[n]
+        if l2_n < bestl2[n] and compare(logits_n, y_n, is_logits=True):
+          bestl2[n] = l2_n
+          bestscore[n] = torch.argmax(logits_n)
```

Comment:
Just a suggestion (not sure how much faster it would be):  Maybe precompute and store the predictions corresponding to `logits_n`, instead of doing argmax calls individually for each batch element (since it's presumably on the GPU, it could benefit from speedups)

Reply:
Pre-computing ```torch.argmax(logits_n)``` outside this ```for``` loop probably won't help since ```torch.argmax(logits_n)``` may or may not be called for each ```n``` so pre-computing for all ```n``` might be a waste.

On the other hand, since ```l2_n < o_bestl2[n]``` being true implies ```l2_n < bestl2[n]``` being true, it makes sense to streamline the implementation there. A new commit has been submitted. Please let me know if it LGTY. Thanks.

Reply:
Yeah, that's a fair point!